### ISSUE TITLE
denying connections before they're established

### ISSUE THREAD

**Issue by pgte** [libp2p/js-libp2p-connection-manager#2]
Allow hooks for denying connections before they are established.

First (and wisely)  requested by @kumavis [here](https://github.com/libp2p/js-libp2p/issues/175).

**Comment by pgte** [comment]
To deny more connections I can think of these 2 options:

1. Close the transport listener so we don't even begin accepting more connections. Reopen once we can accept more connections. I think this needs hooks in `libp2p-switch`, which in turn needs changes in the transport interface and all transport implementations.
2.  When receiving a new raw connection, `libp2p-switch` should emit a `connect` event that is cancelable. If event is canceled, close the connection before doing the multiplex->secio-> multiplex dance.

Option 1 looks more efficient, but harder to implement due to changes.  @diasdavid Do you have an opinion on this?

**Comment by pgte** [comment]
Looks like [the interface defined in `interface-transport`](https://github.com/libp2p/interface-transport#api) already allows for closing (`.close(cb)`) and reopen (`.listen(ma, cb)`).

Whether the specific transports can themselves handle recycling listeners (listen, close, listen) is another question. If some of them don't allow this, it should be categorised as a bug, since the interface does not state that this can't happen.

A problem could exist in specific transports that also do relaying, like websockets-star. If I understand correctly, when closing a listener, the connection to the signaling web socket server is closed. Since this connection may contain internal streams with other peers, besides not accepting new incoming dials, this would also close the existing connections.

If this is true, I think option 2 would work best. If, for instance, [the `_incoming dial` method](https://github.com/libp2p/js-libp2p-websocket-star/blob/master/src/listener.js#L179-L197) would emit a cancelable event, this connection could be closed on the same event loop.

Perhaps @dryajov could shed some light on this?

**Comment by pgte** [comment]
Also, @mkg20001, you have any insight / opinion on this?

**Comment by mkg20001** [comment]
Quick hack for wrtc-star would be to just don't accept webrtc offers after a threshold is reached.

For the long term solution: Maybe instead of closing the listener there could simply be new functions like ".startReject()" and ".startAccept()" (for example this would tell wrtc-star to start/stop accepting offers, yet this wouldn't require reconnecting to the wrtc-star server everytime)

**Comment by dryajov** [comment]
After taking a closer look at the readme, I can see we're limiting the amount of peers being connected to, which can be either over a raw connection or a muxed connection in the case of circuit (and ws-star currently).

> Whether the specific transports can themselves handle recycling listeners (listen, close, listen) is another question. If some of them don't allow this, it should be categorised as a bug, since the interface does not state that this can't happen.

Agreed, we should make the transports "reentrant", so that we can selectively start/stop them.

**Comment by pgte** [comment]
@dryajob thank you for your insights!

The goal here is to save resources. An active connection to a peer consumes memory and CPU (varying on connection activity), which may still be relevant in mobile contexts even if we're relaying through websockets. And in JS, memory pressure equals CPU overhead because of GC. Also, webrct is specially taxing, but I think we could solve this in general independently of transport.

So, even if the p2p connections are relayed and muxed, a connection to a new peer still induces a considerable overhead (a lot of time spent on crypto, etc), which is still taxing on the peer and the relay server.

I think we should, in the API, distinguish between closing the transport (`.close()`) and accepting/rejecting more connections.

For this, I really like your idea of, much like TCP, being able to specify a `maxConnections` option in the transport. This would be a part of the interface, implemented in each transport.

@dryajov thoughts?

**Comment by pgte** [comment]
Of course, having a `maxConnections` option limits some applicability here, since it's a static value, and sometimes we may need to stop accepting connections if some measure (like event loop delay) exceed the threshold.

**Comment by dryajov** [comment]
Sorry, totally missed your reply!

> Of course, having a maxConnections option limits some applicability here, since it's a static value, and sometimes we may need to stop accepting connections if some measure (like event loop delay) exceed the threshold.

I agree, we still need the ability to stop accepting connections on demand - maybe we need a `.start` and `.stop` methods in addition to `listen` and `close`? The new methods will stop accepting connections on demand (even if rejecting on arrival instead of letting the runtime handle it as in the case of `maxConnections`?)

---

There is one more thing thats bugging me (perhaps it even needs it's own issue?), which is, there are some connections that are more important than others. For example, dropping some connections that circuit (`HOP`) is using, can drop a bunch of multiplexed connections that rely on it, how would the connection manager handle it? Do we need to introduce a notion of priority, or just increase counts for that transport to be high enough to prevent premature disconnections?

**Comment by dryajov** [comment]
Sure thing!

TL;DR

- Relay uses a muxed connection to bridge multiple nodes
  - a dialer might have several muxed connections going to a relay over another muxed connection
  - a listener might have many muxed connections coming from a relay point over another muxed connection. 

Dropping the underlying connection to the relay or from the relay (raw or muxed) will drop all the other connections that are established on top of it.

---

So, circuit has three parts
- Dialer
    - dialer that uses libp2p-switch to dial out
- Listener
    - registers a `multistream` handler to listen for incoming connections over circuit
- HOP
    - the actual relay that does the circuiting

Both the dialer and listener are borrowed from the `interface-transports` interface, so in that regard, is just like any other transport, but instead of using TCP or WebSockets directly, to dial out, it uses the `libp2p-switch` itself to dial. The `switch` will use whatever existing connection it has, or establish a new one to dial to the `HOP` node. The `HOP` node again, uses the switch to dial the relayed node.

All dialing happens over muxed connections, and once the connection between the two relayed points is established, it gets encrypted and muxed once again.

Borrowing this from the circuit tutorial, this is more or less how the flow looks:

1. `Node A` tries to connect to `Node B` over one of its known addresses
2. Connection fails because of firewall/NAT/incompatible transports/etc...
3. Both `Node A` and `Node B` know of the same relay - `Relay`
4. `Node A` falls back to dialing over `Relay` to `Node B` using its `'/p2p-circuit'` address, which involves:
   1. `Node A` sends a `HOP` request to `Relay`
   2. `Relay` extracts the destination address, figures out that a circuit to `Node B` is being requested
   3. `Relay` sends a `STOP` request to `Node B`
   4. `Node B` responds with a `SUCCESS` message
   5. `Relay` proceed to create a circuit over the two nodes
5. `Node A` and `Node B` are now connected over `Relay`

The reason why, dropping a connection might drop a bunch of circuited connections, is because that underlying connection can be to a `HOP` node, or in the case of the `HOP` node itself, it can be to a listener (or many listeners). There are a few cases were this happens:

- `HOP` to listener (`STOP`) node
   - i.e. a browser/NATed node with popular content
   - "private services" that only some nodes can access over a relay
- Dialer to `HOP` node 
    - a node that dials out to many nodes over a relay
       - browser to browser (without WebRTC)
       - esoteric transports that only some of the nodes speak (BLE, etc...)

**Comment by pgte** [comment]
@dryajov thank you, that was very clear to me!

Looks to me that a connection to a HOP node should be valued by the peer itself + the values of all the peers we are connected to through it.
To me, it feels like this would be the "true" value measure of a connection to a HOP peer, since disconnecting it would remove all that value.
@dryajov Does this make sense and would this solve this issue?

**Comment by pgte** [comment]
@dryajov also, I guess this would have to be applied recursively, because it looks to me like it's possible to have infinite nesting of HOP connections, or no?

**Comment by dryajov** [comment]
> Looks to me that a connection to a HOP node should be valued by the peer itself + the values of all the peers we are connected to through it.

That sounds correct to me too. 

> I guess this would have to be applied recursively, because it looks to me like it's possible to have infinite nesting of HOP connections, or no?

Hmm - yes. But multi hop dialing is not yet here, so we can probably postpone implementing it for now, tho the reasoning is correct 👍.

---

I also want to bring up one more thing - the go implementation has a concept of tagging peers, maybe we can borrow that as well... 

Basically, as far as I can tell, tagging boils down to being able to set an arbitrary value for a peer - _all tags contribute to the peers value - https://github.com/libp2p/go-libp2p-connmgr/blob/master/connmgr.go#L136...L148_, the peers with the lowest value then get disconnected - https://github.com/libp2p/go-libp2p-connmgr/blob/master/connmgr.go#L61...L108. Peers with higher importance, get a higher number https://github.com/libp2p/go-libp2p-circuit/pull/20 and https://github.com/libp2p/go-libp2p-kad-dht/pull/95.

IMO, this is a reasonable and simple way of prioritizing certain peers over others...

Maybe we set the default to 0.5 and _go_ from there, tho if we're going to do this, it might make sense to switch to whole numbers, since it makes things easier(?) to handle...

**Comment by pgte** [comment]
@dryajov thanks for showing that, I wasn't aware that go managed it like this (I thought it disconnected peers by order of how old the connection was).

Currently, here we have a global value per peer, which defaults to 1 and can be overridden by the app.
I guess that you're catering for the scenario where there may be multiple applications using the same IPFS node, and that the tag system may be used to represent the value for each application?

**Comment by dryajov** [comment]
> Currently, here we have a global value per peer, which defaults to 1 and can be overridden by the app.

Yep, we're pretty close in essence to the Go approach.

> I guess that you're catering for the scenario where there may be multiple applications using the same IPFS node

I believe tagging can be applied to that as well (tho, I'm not sure what was the initial use case that Go was going after). My only aim with this (which I believe can be achieved without tagging), is to _allow peers to be marked as having higher priority than another peer_.

i.e.: We can allow a service/app/transport to increase a peer's `value` based on some criteria, that when an event gets triggered, allows higher valued peers to be evicted after lower valued peers.

In this case, our concept of peer value works just as well and if we can allow for it to be adjusted arbitrarily, it'll essentially be the same thing without namespaces/tags.

// cc @Stebalien @whyrusleeping @vyzo

**Comment by whyrusleeping** [comment]
On the go side we don't currently have the facilities in place to close out connections before they are established. The main issue with that is that you have to do a full secio handshake anyways before you can know which peer it is, and you don't want to blanket ignore *all* new connections

**Comment by pgte** [comment]
Opened "peer value should be HOP-aware" issue: https://github.com/libp2p/js-libp2p-connection-manager/issues/10

**Comment by pgte** [comment]
Moving the interface discussion to here: https://github.com/libp2p/interface-transport/issues/34

**Comment by pgte** [comment]
Now tracking here: https://github.com/libp2p/js-libp2p-switch/issues/251

**Comment by diasdavid** [comment]
@pgte is there value to keep this issue open?

**Comment by pgte** [comment]
@diasdavid yes, I think so, since it's the job of the connection manager to tell the transports whether they should be accepting more connections or not.

**Comment by mkg20001** [comment]
I think making .start()/.stop() not close existing connections is a bad idea because when the libp2p instance gets stopped it might leave unclosed connections around.
Instead there should be a dedicated pair of functions .accept()/.reject() that change whether or not a transport will accept connections and .stop() should close all connections

**Comment by pgte** [comment]
@mkg20001 I think that the transport interface discussion is happening here: https://github.com/libp2p/interface-transport/issues/34
Could you post this there?

**Comment by pgte** [comment]
@mkg20001 I see the issue I pointed you to is closed. Perhaps create a new issue there?

**Comment by mkg20001** [comment]
@pgte Created the issue: https://github.com/libp2p/interface-transport/issues/39
